Make all-in-one.yaml file independant of sampling-strategies.json file

adityachopra29 opened the pull request:
## Which problem is this PR solving?
- fix #6338 

## Description of the changes
-  Removed the hardcoded path in the default configuration (all-in-one.yaml).
- updated tests accordingly

## How was this change tested?
- Since some of the tests are using hardcoded services in ```sampling-strategies.json``` (the service "foo").
https://github.com/jaegertracing/jaeger/blob/b02900cd99b38a79f70d9ba14f631307ded84764/cmd/jaeger/internal/extension/remotesampling/extension_test.go#L83
the ```sampling-strategies.json``` file could not be removed
- Also, right now there are some tests, namely:
1.) TestServerHTTP_TracesRequest,
2.) all 3 tests in cmd/query/app
which are not passing even on the main branch itself (according to me, otherwise I have made some mistake), and hence these same tests are not passing after I made my changes (my changes should not affect them anyway). 


## Checklist
- [X] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [X] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [X] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `npm run lint` and `npm run test`

yurishkuro commented on cmd/jaeger/internal/extension/remotesampling/config_test.go:
```diff
@@ -17,11 +17,6 @@ func Test_Validate(t *testing.T) {
 		config      *Config
 		expectedErr string
 	}{
-		{
```
this is still a valid test

yurishkuro commented on cmd/jaeger/internal/extension/remotesampling/extension_test.go:
```diff
@@ -84,7 +83,7 @@ func makeRemoteSamplingExtension(t *testing.T, cfg component.Config) component.H
 	factory := NewFactory()
 	cfg := factory.CreateDefaultConfig().(*Config)
-	cfg.File.Path = filepath.Join("..", "..", "..", "sampling-strategies.json")
+	cfg.File = nil
```
why do you need to change this test?

adityachopra29 replied:
Apologies for the delay.
I was trying to make the the tests independant of the sampling-strategies file. And so I was giving a null value to the input input to prompt default behavior. The test was working either ways(input was file based in both cases finally).
Still, I have removed the changes now.

yurishkuro reviewed (commented):
* you are working off a stale `main`, please rebase. There should be no changes to submodules in your PR
* you have not made any changes to the logic, so why are you sending this for review?
* you need to test that sampling endpoint returns expected stratery
```
$ curl "http://localhost:5778/?service=x"
{"strategyType":0,"probabilisticSampling":{"samplingRate":1}}%
```

adityachopra29 commented:
Hi @yurishkuro,
Kindly review the PR now.
- the output is coming correct testing the endpoint 
``` 
$ curl "http://localhost:5778/?service=x"
{"strategyType":0,"probabilisticSampling":{"samplingRate":0.001}}%
```
- I have updated the base main branch
- no other changes were suggested in logic, hence rest code is same

yurishkuro commented on cmd/jaeger/internal/extension/remotesampling/config_test.go:
```diff
@@ -44,13 +44,6 @@ func Test_Validate(t *testing.T) {
-			config: &Config{
-				File: &FileConfig{Path: ""},
-			},
-			expectedErr: "File.Path: non zero value required",
```
keep this test but with `expectedErr=""`

yurishkuro commented:
```
$ docker run -p 5778:5778 jaegertracing/jaeger:2.1.0

$ curl "http://localhost:5778/?service=x"
{"strategyType":0,"probabilisticSampling":{"samplingRate":1}}%
```

The current default behavior is to return probability 1.0, not 0.001.

There is a hardcoded constant `defaultSamplingProbability = 0.001`, but we don't want to change it because jaeger-v1 should still be using that, so we need to make that value configurable in the static/provider

yurishkuro commented:
make sure to pull from remote branch before you continue work, to pick up the changes I made

adityachopra29 commented:
Okay. So the initial issue regarding the dependancy of sampling-strategies.json file in all-in-one.yaml is resolved right? This is a new issue which needs to be solved? 
To make ```SamplingRate``` configurable in static/provider, keeping ```defaultSamplingProbability``` const?
If yes, then I plan on implementing this by creating a factor (with default value 1000), which I will multiply ```defaultSamplingProbability``` with when returning ```SamplingRate```. Am I following the right approach?

yurishkuro commented:
The original issue is not resolved since these changes return different results from what v2.1.0 Jaeger container returns.

There is no need for any multiplier, the default probability should be made parameterizable via the provider API so that we can pass it from jaeger-v2 sampling extension.

adityachopra29 commented:
@yurishkuro Please have a look once now. 
- A parameter for DefaultSamplingProbability has been created which can be filled using a flag, or using the yaml file input default_sampling_probability.
- A flag called "sampling.stategies-default-sampling-probability" has been created for the same
- If no input is given for DefaultSamplingProbability, it defaults to "defaultDefaultSamplingProbability" = 0.001. (which was originally used)
- DefaultSamplingProbability has been given value 0.001 in all-in-one.yaml (to give the default value as before)
- In all the tests given in the provider_test.go file, initially they used defaultSamplingProbability = 0.001. Since now this parameterized, I have used the value of defaultDefaultSamplingProbability = 0.001 instead now to keep the default behavior.
- Since no changes have been made to the original defaultSamplingProbability variable, the output still comes as follows: 

```
 curl "http://localhost:5778/?service=x"
{"strategyType":0,"probabilisticSampling":{"samplingRate":0.001}}%  
```
Instead of 
```
{"strategyType":0,"probabilisticSampling":{"samplingRate":1}}%  
```
whch you wanted. I wanted to confirm what change I have to do to set this.

yurishkuro commented on cmd/jaeger/internal/extension/remotesampling/config.go:
```diff
@@ -104,6 +107,10 @@ func (cfg *Config) Validate() error {
 		return errNegativeInterval
 	}
 
+	if cfg.File != nil && (cfg.File.DefaultSamplingProbability > 1 || cfg.File.DefaultSamplingProbability < 0) {
```
I think you should be able to achieve similar validation by adding a `valid` tag to the field, perhaps `valid:range(0, 1)`

yurishkuro commented on cmd/jaeger/internal/extension/remotesampling/config.go:
```diff
@@ -47,9 +48,11 @@ type Config struct {
 	// ReloadInterval is the time interval to check and reload sampling strategies file
 	ReloadInterval time.Duration `mapstructure:"reload_interval"`
+	// DefaultSamplingProbability is the sampling probability used by the Strategy Store for static sampling
+	DefaultSamplingProbability float64 `mapstructure:"default_sampling_probability"`
```
you need to assign default value to this in factory/createDefaultConfig

yurishkuro commented on cmd/jaeger/internal/all-in-one.yaml:
```diff
@@ -30,8 +30,9 @@ extensions:
-      path: ./cmd/jaeger/sampling-strategies.json
+      path:
       reload_interval: 1s
+      default_sampling_probability: 0.001
```
why not 1.0? I think this would be equivalent to the previous config where the stategies file was specified

yurishkuro reviewed (commented):
please remove idl submodule change

yurishkuro commented on plugin/sampling/strategyprovider/static/options.go:
```diff
@@ -12,9 +12,10 @@ import (
+	samplingStrategiesFile                       = "sampling.strategies-file"
+	samplingStrategiesReloadInterval             = "sampling.strategies-reload-interval"
+	samplingStrategiesBugfix5270                 = "sampling.strategies.bugfix-5270"
+	samplingStrategiesDefaultSamplingProbability = "sampling.stategies-default-sampling-probability"
```
```suggestion
	samplingStrategiesDefaultSamplingProbability = "sampling.default-sampling-probability"
```

yurishkuro commented on plugin/sampling/strategyprovider/static/options.go:
```diff
@@ -27,19 +28,23 @@ type Options struct {
 	flagSet.Duration(samplingStrategiesReloadInterval, 0, "Reload interval to check and reload sampling strategies file. Zero value means no reloading")
 	flagSet.String(samplingStrategiesFile, "", "The path for the sampling strategies file in JSON format. See sampling documentation to see format of the file")
 	flagSet.Bool(samplingStrategiesBugfix5270, true, "Include default operation level strategies for Ratesampling type service level strategy. Cf. https://github.com/jaegertracing/jaeger/issues/5270")
+	flagSet.Float64(samplingStrategiesDefaultSamplingProbability, defaultDefaultSamplingProbability, "Sampling probability used by the Strategy Store for static sampling. Value must be between 0 and 1.")
```
> defaultDefaultSamplingProbability

default-default?

yurishkuro commented on plugin/sampling/strategyprovider/static/provider.go:
```diff
@@ -71,13 +71,13 @@ func NewProvider(options Options, logger *zap.Logger) (ss.Provider, error) {
 
 	if options.ReloadInterval > 0 {
-		go h.autoUpdateStrategies(ctx, options.ReloadInterval, loadFn)
+		go h.autoUpdateStrategies(ctx, options.ReloadInterval, loadFn, options)
```
* if you are passing options no need to pass options.ReloadInterval
* do we need to pass either of those? It seems `h` already has access to `h.options`

adityachopra29 replied:
yes it seems we dont need to pass them. Removed now.

yurishkuro reviewed (commented):
on the right track

adityachopra29 commented:
@yurishkuro I have made the changes. Please have a look.
Also, I found that defaultSamplingProbability was not being used anywhere, except in the provider tests(which I replaced with DefaultDefaultSamplingProbability,as you can see. So I removed it.

yurishkuro commented:
please ensure all commits a signed.

adityachopra29 commented:
@yurishkuro  done .... some old commits were missing the signoff

yurishkuro commented on plugin/sampling/strategyprovider/static/constants.go:
```diff
@@ -16,14 +16,14 @@ const (
-	defaultSamplingProbability = 0.001
+	// DefaultDefaultSamplingProbability is the default value for  "DefaultSamplingProbability"
+	// used by the Strategy Store in case no DefaultSamplingProbability is defined
+	DefaultDefaultSamplingProbability = 0.001
```
```suggestion
	DefaultSamplingProbability = 0.001
```

adityachopra29 replied:
Done.

yurishkuro commented on cmd/jaeger/internal/all-in-one.yaml:
```diff
@@ -30,7 +30,8 @@ extensions:
     file:
-      path: ./cmd/jaeger/sampling-strategies.json
+      path:
+      default_sampling_probability: 1
```
Would it make sense as a field at the top level? Right now we have `file.default_sampling_probability` and `adaptive.initial_sampling_probability`, but their meaning is very similar - which probability to return in the absence of any better information.
